Support Backup and Restore for GCP

ketakig12 opened the pull request:
- Introduced a new class GcpClient.py
- Supported all basic functions same as other clients
- Added credentials/config params needed for GCP
- Added python api library for compute, cloud storage library for storage and auth libraries for GCP authentication as dependencies

For more details please look into [issues-7](https://github.com/cloudfoundry-incubator/service-fabrik-backup-restore/issues/7)

kamath-prasad commented on lib/clients/GcpClient.py:
```diff
@@ -0,0 +1,555 @@
+                        for instance in instances_scoped_list['instances']:
+                            if instance['name'] == instance_id:
+                                # Assuming the last part of the url is zone-name
+                                return instance['zone'].rsplit('/', 1)[1]
```
do you want to add a null check before the split ? Or is it always confirmed that the value for instance['zone'] would be present ?

ketakig12 replied:
It should ideally be there. But in case not there it will be caught in exception and zone will be none, which will make the code exit at [condition](https://github.com/cloudfoundry-incubator/service-fabrik-backup-restore/pull/8/files#diff-54e75b3eae7ffacb806944d3cc96d724R45)

muralisc commented on lib/clients/GcpClient.py:
```diff
@@ -0,0 +1,555 @@
+                response = request.execute()
+                for zone, instances_scoped_list in response['items'].items():
+                    if 'warning' in instances_scoped_list:
+                        pass
```
Extra if check doing nothing. 
Can can have only one if check which checks the case below?

ketakig12 replied:
Yes. will make the change.

muralisc commented on lib/clients/GcpClient.py:
```diff
@@ -0,0 +1,555 @@
+        pass
+    '''
+
+    def wait_for_operation(self, operation_id, zonal_operation):
```
Better name ? `get_operation_status`
No actual waiting is done in this function.

ketakig12 replied:
Yes, will change it to 'get_operation_status'.

ketakig12 commented on lib/clients/BaseClient.py:
```diff
@@ -135,25 +135,29 @@ def _add_volume_device(self, volume_id, device):
 
     def _remove_volume_device(self, volume_id):
-        self.__devices.pop(volume_id, None)
+        if volume_id in self.__devices:
```
I was getting an exception/error when the item was not found.

muralisc replied:
are you sure it was `_remove_volume_device` ? it should be the case for `_remvoe_snapshot`
remove volume uses dict and remove snapshot used list
sample expt in Python
```
>>> var={ "value1": 999 }
>>> var.pop('value2')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
KeyError: 'value2'
>>> var.pop('value2',None)
>>> # no error thrown
```

ketakig12 replied:
Yes, you are right. As both the functions ```_remove_volume_device``` and ```_remove_attachment``` are getting called in  [_delete_attachment](https://github.com/ketakig12/service-fabrik-backup-restore/blob/02a1226f515de458156b6b2749aaa000aa22a3e0/lib/clients/GcpClient.py#L434-L435), I misunderstood the error and added the check in both the functions. Changed it for ```_remove_volume_device```

subhankarc commented on lib/clients/GcpClient.py:
```diff
@@ -0,0 +1,553 @@
+                        for instance in instances_scoped_list['instances']:
+                            if instance['name'] == instance_id:
+                                # Assuming the last part of the url is zone-name
+                                return instance['zone'].rsplit('/', 1)[1]
```
Out of curiosity, how does the zone attribute look like? Wondering if there are exceptions to the rule!

ketakig12 replied:
Example zone attribute looks like:
``` "zone": "https://www.googleapis.com/compute/v1/projects/project_name/zones/europe-west1-b" ```
Specification of the api used here is at: [instances/aggregatedList](https://cloud.google.com/compute/docs/reference/latest/instances/aggregatedList). Here **"instances"** details under "**key**" refer to [instances#resource](https://cloud.google.com/compute/docs/reference/latest/instances#resource). In this, zone refers to 'URL of the zone where the instance resides.' and this URL specification can be found in [zones/get](https://cloud.google.com/compute/docs/reference/latest/zones/get).

sauravmndl commented on lib/clients/GcpClient.py:
```diff
@@ -0,0 +1,553 @@
+        self.compute_api_name = 'compute'
+        self.compute_api_version = 'v1'
+        self.snapshot_prefix = 'sf-snapshot'
+        self.disk_prefix = 'sf-disk'
```
common prefixes move to BaseClient.py

ketakig12 replied:
Sure, will change.

sauravmndl commented on lib/clients/GcpClient.py:
```diff
@@ -0,0 +1,553 @@
+            result = self.compute_client.globalOperations().get(
+                project=self.project_id,
+                operation=operation_id).execute()
+        return result['status']
```
is the variable 'result' outside of if..else? Just check once.

ketakig12 replied:
Yes, it is accessible outside if..else as well. Still will declare it as None outside. Also, improving the function for better error handling.

sauravmndl commented on lib/clients/GcpClient.py:
```diff
@@ -0,0 +1,553 @@
+    def _download_from_blobstore_and_pipe_to_process(self, process, blob_to_download_name, segment_size):
+        # TODO Currently not used for GCP. Implement if needed.
+        pass
+    '''
```
If not required, you can remove it.

ketakig12 replied:
Sure, will remove.

sauravmndl commented on lib/clients/GcpClient.py:
```diff
@@ -0,0 +1,553 @@
+                           If you wish to use resumable upload, pass chunk_size param to this function.
+                           This must be a multiple of 256 KB per the API specification.
+        """
+        log_prefix = '[Google Cloud Storage] [UPLOAD]'
```
keep log prefixes consistent. use 'GCP' or 'Google Cloud Storage'

sauravmndl commented on lib/clients/GcpClient.py:
```diff
@@ -0,0 +1,553 @@
+    def _create_volume(self, size, snapshot_id=None):
+        log_prefix = '[VOLUME] [CREATE]'
+        volume = None
+
```
Remove extra line.

ketakig12 replied:
Okay, removing from other places as well.

sauravmndl commented on lib/clients/GcpClient.py:
```diff
@@ -0,0 +1,553 @@
+                       None,
+                       disk_creation_operation['name'], True)
+
+            volume = self.get_volume(disk_name)
```
If the volume/disk details are available from **disk_creation_operation** then avoid extra IaaS call made inside **self.get_volume(disk_name)**.

ketakig12 replied:
disks.insert returns operation object which only provides information on the status of the operation. See [disk/insert](https://cloud.google.com/compute/docs/reference/rest/v1/disks/insert). Hence, we cannot avoid the extra call.

sauravmndl commented on lib/clients/GcpClient.py:
```diff
@@ -0,0 +1,553 @@
+                       None,
+                       disk_deletion_operation['name'], True)
+
+            volume_exists = self.volume_exists(volume_id)
```
If delete operation returns success. do you still need to cross check for existence of disk?

ketakig12 replied:
Ideally, not needed. But added it as an extra precaution as we might have dangling volumes if not deleted properly.

sauravmndl commented on lib/clients/GcpClient.py:
```diff
@@ -0,0 +1,553 @@
+                    'Attached volume-id={}, device={}'.format(volume_id, device))
+                self._add_volume_device(volume_id, device)
+            else:
+                raise Exception()
```
Put some logs/ messages inside the Exception()

ketakig12 replied:
Sure, will add.

sauravmndl commented on lib/clients/GcpClient.py:
```diff
@@ -0,0 +1,553 @@
+
+            # Here volume_id is the device name.
+            # Raise exception if device returned in None,
+            # as it might mean that disk was not attached properly
```
If the disk is not attached, wouldn't it throw exception? Do we need to cross check for device?

ketakig12 replied:
Ideally, not needed. But added it as an extra check.

ketakig12 commented on lib/clients/GcpClient.py:
```diff
@@ -0,0 +1,553 @@
+from google.oauth2 import service_account
+from google.cloud import storage
+from google.cloud.storage import Blob
+from .BaseClient import BaseClient
```
Sorry, will remove.

ketakig12 commented on lib/clients/GcpClient.py:
```diff
@@ -0,0 +1,553 @@
+class GcpClient(BaseClient):
+    def __init__(self, operation_name, configuration, directory_persistent, directory_work_list, poll_delay_time,
+                 poll_maximum_time):
+        super(GcpClient, self).__init__(operation_name, configuration, directory_persistent, directory_work_list,
```
`super().method()` also does the same thing as `super(C, self).method(arg)`. Since in other IaaS clients as well, it is specified in the same way, it will be better to keep it same way everywhere.

jagadish-kb replied:
ok

ketakig12 commented on lib/clients/GcpClient.py:
```diff
@@ -0,0 +1,553 @@
+                                        poll_delay_time, poll_maximum_time)
+
+        self.__gcpCredentials = json.loads(configuration['credentials'])
+        self.project_id = configuration['projectId']
```
AFAIU, double underscore will mangle the attribute names of a class to avoid conflicts of attribute names between classes, see [understanding-the-underscore-of-python](https://hackernoon.com/understanding-the-underscore-of-python-309d1a029edc). So, it is not exactly meant for private access, but can be used so.

jagadish-kb replied:
ok

ketakig12 commented on lib/clients/GcpClient.py:
```diff
@@ -0,0 +1,553 @@
+
+        # +-> Get the availability zone of the instance
+        self.availability_zone = self._get_availability_zone_of_server(
+            configuration['instance_id'])
```
Instance Id is passed from broker in vms.iaas_vm_metadata.vm_id. See [#backup](https://cloudfoundry-incubator.github.io/service-fabrik-broker/api/agent_v1.1.html#backup). And I think broker in turn gets it from BOSH, [DirectorManager.js](https://github.com/cloudfoundry-incubator/service-fabrik-broker/blob/b99ec4bb1821b9bbc4ce50c4ae2efd96955d34c5/lib/fabrik/DirectorManager.js#L596) , [BoshDirectorClient.js](https://github.com/cloudfoundry-incubator/service-fabrik-broker/blob/28ded14aa0436f53c7d1dea25d7bf221f5fe39cc/lib/bosh/BoshDirectorClient.js#L396). Please correct if wrong.

jagadish-kb replied:
At least an analysis of whether instance id can be retrieved locally will be helpful.
Documentation: https://cloud.google.com/compute/docs/storing-retrieving-metadata
https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-metadata.html#instancedata-data-retrieval

ketakig12 replied:
Yes, looks name is the VM id and can be retrieved with 
`
curl "http://metadata.google.internal/computeMetadata/v1/instance/name" -H "Metadata-Flavor: Google"
`
So, if we decide to go with this approach, we can.

ketakig12 commented:
Review comments changes:
- Improvement of get_operation and additional checks for snapshot/volume to be in READY state
- Moved prefix to BaseClient
- Removed unnecessary method impl _download_from_blobstore_and_pipe_to_process
- Removed extra lines
- Added logs/messages inside the Exception()
- Removed duplicate BaseClient import

ketakig12 commented:
Unit tests will be added as part of #13

ketakig12 commented on lib/clients/GcpClient.py:
```diff
@@ -230,16 +229,21 @@ def _create_snapshot(self, volume_id):
-            self.logger.info('{} SUCCESS: snapshot-id={}, volume-id={} with tags {}'.format(
-                log_prefix, snapshot.id, volume_id, self.tags))
-
+            if snapshot.status == 'READY':
```
I could not find any enum in the directory. See even examples contain the string status [create_instance.py](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/a2d8238250e6c18e25b5aa3123a7b6ec367613f7/compute/api/create_instance.py#L136)
